File: CLI_Task_Timer/task_timer.js

const fs = require('fs'); // load the file system module to handle file operations
const inquirer = require('inquirer'); // load the inquirer module for user prompts

const taskDataPath = './tasks.json'; // define the file path for storing tasks

// load or initialize tasks
const loadTasks = () => {
    if (fs.existsSync(taskDataPath)) { // check if the task data file exists
        return JSON.parse(fs.readFileSync(taskDataPath, 'utf8')); // read and parse the task data file
    }
    return {}; // return an empty object if no file exists
};

// save tasks to file
const saveTasks = (tasks) => {
    fs.writeFileSync(taskDataPath, JSON.stringify(tasks, null, 2)); // write tasks to the file as formatted json
};

// get current timestamp
const getCurrentTime = () => new Date().toISOString(); // return the current date and time in iso format

// main task logic
const taskManager = {
    start(taskName) { // start a new task
        const tasks = loadTasks(); // load existing tasks
        if (tasks[taskName] && tasks[taskName].status === 'active') { // check if the task is already running
            console.log(`Task "${taskName}" is already running.`); // notify user that the task is active
            return; // exit the function
        }
        tasks[taskName] = { // create or update the task
            startTime: getCurrentTime(), // set the start time
            totalTime: tasks[taskName]?.totalTime || 0, // keep existing total time or set to zero
            status: 'active', // set the task status to active
        };
        saveTasks(tasks); // save the updated tasks to file
        console.log(`Started task "${taskName}".`); // notify user that the task has started
    },
    

    resume(taskName) { // resume a paused task
        const tasks = loadTasks(); // load existing tasks
        if (!tasks[taskName] || tasks[taskName].status !== 'paused') { // check if the task is not paused
            console.log(`Task "${taskName}" is not currently paused.`); // notify user that the task is not paused
            return; // exit the function
        }
        tasks[taskName].startTime = getCurrentTime(); // update the start time
        tasks[taskName].status = 'active'; // set the task status to active
        saveTasks(tasks); // save the updated tasks to file
        console.log(`Resumed task "${taskName}".`); // notify user
    },

    status(taskName) { // check the status of a task
        const tasks = loadTasks(); // load existing tasks
        if (!tasks[taskName]) { // check if the task does not exist
            console.log(`Task "${taskName}" does not exist.`); // notify user that the task does not exist
            return; // exit the function
        }
        console.log(`Task "${taskName}" status: ${tasks[taskName].status}. Total time: ${tasks[taskName].totalTime.toFixed(2)} seconds.`); // display task status
    },

    summary() { // display a summary of all tasks
        const tasks = loadTasks(); // load existing tasks
        console.log('Task Summary:'); // print a header
        Object.entries(tasks).forEach(([taskName, task]) => { // iterate over all tasks
            console.log(`- ${taskName}: ${task.totalTime.toFixed(2)} seconds (${task.status})`); // display task details
        });
    },
};

// cli interface
const mainMenu = async () => { // display the main menu
    const { command } = await inquirer.prompt([ // prompt user for a command
        {
            type: 'list', // use a list prompt
            name: 'command', // name the prompt result
            message: 'What would you like to do?', // display a message to the user
            choices: [ // define the available choices
                'Start a task',
                'Stop a task',
                'Pause a task',
                'Resume a task',
                'Check task status',
                'View summary',
                'Exit',
            ],
        },
    ]);

    if (command === 'Exit') { // check if the user chose to exit
        console.log('Goodbye!'); // say goodbye
        return; // exit the program
    }

    if (command === 'View summary') { // check if the user chose to view the summary
        taskManager.summary(); // display the summary
    } else { // handle all other commands
        const { taskName } = await inquirer.prompt([ // prompt the user for a task name
            {
                type: 'input', // use an input prompt
                name: 'taskName', // name the prompt result
                message: 'Enter the task name:', // ask the user for the task name
            },
        ]);
        switch (command) { // handle the command
            case 'Start a task': // if the command is to start a task
                taskManager.start(taskName); // start the task
                break;
            case 'Stop a task': // if the command is to stop a task
                taskManager.stop(taskName); // stop the task
                break;
            case 'Pause a task': // if the command is to pause a task
                taskManager.pause(taskName); // pause the task
                break;
            case 'Resume a task': // if the command is to resume a task
                taskManager.resume(taskName); // resume the task
                break;
            case 'Check task status': // if the command is to check task status
                taskManager.status(taskName); // check the task status
                break;
        }
    }

    await mainMenu(); // loop back to menu
};

// run the app
mainMenu(); // start the main menu

File: task_timer.py
#!/usr/bin/env python3

"""
Task Timer CLI
--------------
A terminal-based program to track, pause, resume, stop, and report on tasks.
Data is stored in a local JSON file named 'tasks.json'.

Usage:
  python task_timer.py start <task_name>
  python task_timer.py stop <task_name>
  python task_timer.py pause <task_name>
  python task_timer.py resume <task_name>
  python task_timer.py status <task_name>
  python task_timer.py summary

Examples:
  python task_timer.py start "My Task"
  python task_timer.py status "My Task"
  python task_timer.py pause "My Task"
  python task_timer.py resume "My Task"
  python task_timer.py stop "My Task"
  python task_timer.py summary
"""

import sys
import json
import os
from datetime import datetime

TASK_FILE = 'tasks.json'

def load_tasks():
    """
    Load tasks from the JSON file, or return an empty dictionary if none exists.
    """
    if os.path.isfile(TASK_FILE):
        with open(TASK_FILE, 'r') as f:
            return json.load(f)
    return {}

def save_tasks(tasks):
    """
    Save the tasks dictionary to the JSON file in a formatted way.
    """
    with open(TASK_FILE, 'w') as f:
        json.dump(tasks, f, indent=2)

def get_current_time():
    """
    Return the current time as an ISO 8601 string.
    """
    return datetime.now().isoformat()

def get_display_time(task_data):
    """
    Return the displayed time for a task:
    - If 'active', add live elapsed time to the 'totalTime'.
    - If 'paused' or 'completed', just return the 'totalTime'.
    """
    status = task_data['status']
    total_time = task_data['totalTime']
    if status == 'active':
        # Calculate how much time has elapsed since last start
        start_time = datetime.fromisoformat(task_data['startTime'])
        elapsed = (datetime.now() - start_time).total_seconds()
        return total_time + elapsed
    else:
        # Paused or completed; just use totalTime
        return total_time

def start(task_name):
    """
    Start a new task (or resume an existing completed one). Updates the status to 'active'.
    """
    tasks = load_tasks()

    # If a task is already active, warn the user
    if task_name in tasks and tasks[task_name]['status'] == 'active':
        print(f"Task \"{task_name}\" is already running.")
        return

    tasks[task_name] = {
        'startTime': get_current_time(),
        'totalTime': tasks.get(task_name, {}).get('totalTime', 0.0),
        'status': 'active'
    }
    save_tasks(tasks)
    print(f"Started task \"{task_name}\".")

def stop(task_name):
    """
    Stop an active task, mark it as completed, and update 'totalTime'.
    """
    tasks = load_tasks()

    if task_name not in tasks or tasks[task_name]['status'] != 'active':
        print(f"Task \"{task_name}\" is not currently active.")
        return

    start_time = datetime.fromisoformat(tasks[task_name]['startTime'])
    elapsed_time = (datetime.now() - start_time).total_seconds()
    tasks[task_name]['totalTime'] += elapsed_time
    tasks[task_name]['status'] = 'completed'
    save_tasks(tasks)
    print(f"Stopped task \"{task_name}\". Total time: {tasks[task_name]['totalTime']:.2f} seconds.")

def status(task_name):
    """
    Print the status of a specific task, including up-to-date total time if it's active.
    """
    tasks = load_tasks()
    if task_name not in tasks:
        print(f"Task \"{task_name}\" does not exist.")
        return

    current_time = get_display_time(tasks[task_name])
    print(f"Task \"{task_name}\" status: {tasks[task_name]['status']}. "
          f"Total time: {current_time:.2f} seconds.")

def summary():
    """
    Display a summary of all tasks, including the updated time if they're active.
    """
    tasks = load_tasks()
    if not tasks:
        print("No tasks found.")
        return

    print("Task Summary:")
    for t_name, t_data in tasks.items():
        display_time = get_display_time(t_data)
        print(f"- {t_name}: {display_time:.2f} seconds ({t_data['status']})")

def usage():
    """
    Print usage instructions.
    """
    print("Usage:")
    print("  python task_timer.py start <task_name>")
    print("  python task_timer.py stop <task_name>")
    print("  python task_timer.py pause <task_name>")
    print("  python task_timer.py resume <task_name>")
    print("  python task_timer.py status <task_name>")
    print("  python task_timer.py summary")
    print("\nExamples:")
    print("  python task_timer.py start \"My Task\"")
    print("  python task_timer.py status \"My Task\"")
    print("  python task_timer.py pause \"My Task\"")
    print("  python task_timer.py resume \"My Task\"")
    print("  python task_timer.py stop \"My Task\"")
    print("  python task_timer.py summary")

def main():
    if len(sys.argv) < 2:
        usage()
        sys.exit(1)

    command = sys.argv[1].lower()

    if command == 'start':
        if len(sys.argv) < 3:
            print("Error: Missing task name.")
            sys.exit(1)
        task_name = " ".join(sys.argv[2:])
        start(task_name)

    elif command == 'stop':
        if len(sys.argv) < 3:
            print("Error: Missing task name.")
            sys.exit(1)
        task_name = " ".join(sys.argv[2:])
        stop(task_name)


    elif command == 'status':
        if len(sys.argv) < 3:
            print("Error: Missing task name.")
            sys.exit(1)
        task_name = " ".join(sys.argv[2:])
        status(task_name)

    elif command == 'summary':
        summary()

    else:
        usage()
        sys.exit(1)

if __name__ == '__main__':
    main()
